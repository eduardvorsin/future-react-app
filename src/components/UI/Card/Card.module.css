.card {
	border-radius: 4px;
	padding: 15px;
	display: flex;
	flex-direction: column;
	justify-content: space-evenly;
	background-color: var(--brand-color-100);
	transition: background-color var(--transition-time) var(--transition-easing);
	position: relative;
}

.card__img-wrapper {
	align-self: center;
	margin-bottom: 30px;
	position: relative;
}

.card__img-wrapper--mature::after {
	content: '18+';
	color: var(--error-color-200);
	font-size: 28px;
	position: absolute;
	top: 0;
	left: 0;
	width: 100%;
	height: 100%;
	background: hsla(0deg, 0%, 100%, 20%);
	backdrop-filter: blur(7px);
	display: flex;
	align-items: center;
	justify-content: center;
}

.card__img-wrapper>a {
	font-size: 0;
	position: absolute;
	top: 0;
	left: 0;
	width: 100%;
	height: 100%;
	z-index: 5;
}

.card__img {
	object-fit: cover;
}

.card__price {
	margin: 0;
	font-weight: 700;
	margin-bottom: 10px;
	font-size: 20px;
	color: var(--text-color-primary);
	transition: color var(--transition-time) var(--transition-easing);
}

.card__content {
	margin-bottom: 20px;
	flex-grow: 1;
}

.card__category {
	margin: 0;
	font-size: 16px;
	margin-bottom: 15px;
	-webkit-line-clamp: 2;
}

.card__category>span {
	font-weight: 700;
}

.card__title {
	margin: 0;
	text-align: left;
	margin-bottom: 15px;
	-webkit-line-clamp: 3;
	font-size: 18px;
	color: var(--text-color-primary);
}

.card__authors {
	margin: 0;
	color: var(--text-color-primary);
	-webkit-line-clamp: 2;
}

.card__category,
.card__title,
.card__authors {
	-webkit-box-orient: vertical;
	display: -webkit-box;
	overflow: hidden;
	text-overflow: ellipsis;
}

.card__more-link {
	justify-content: center;
}

@media(min-width:576px) {
	.card::after {
		content: '';
		bottom: -20px;
		left: 0;
		width: 120%;
		height: 1px;
		background-color: var(--secondary-color-300);
		position: absolute;
	}
}

@media(prefers-color-scheme:dark) {
	.card {
		background-color: var(--secondary-color-200);
		transition: background-color var(--transition-time) var(--transition-easing);
	}

	.card__price {
		color: var(--brand-color-200);
		transition: color var(--transition-time) var(--transition-easing);
	}
}