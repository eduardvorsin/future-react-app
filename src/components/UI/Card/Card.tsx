import React from 'react';
import { Link } from 'react-router-dom';
import Title from '../Title/Title';
import classes from './Card.module.css';
import PlaceholderImage from '../../../assets/images/placeholder128x185-ru.jpeg';

type CardProps = {
  id: string,
  className?: string,
  category?: string,
  title: string,
  src?: string,
  alt: string,
  authors?: string[],
}

const Card: React.FC<CardProps> = ({
  id,
  className,
  category,
  title,
  src,
  alt,
  authors,
}) => {
  const cardClasses = [
    className,
    classes.card,
  ].join(' ');

  const currentAuthors = authors?.join(', ') ?? 'Неизвестно';
  const currentCategory = category ?? 'Незвестно';
  const currentImage = src ?? PlaceholderImage;

  return (
    <article
      className={cardClasses}
    >
      <img
        className={classes.card__img}
        src={currentImage}
        alt={alt}
      >
      </img>
      <p
        className={classes.card__category}
      >
        Категория: {currentCategory}
      </p>
      <Title
        className={classes.card__title}
        level={6}
        component='h3'
      >
        {title}
      </Title>
      <p
        className={classes.card__authors}
      >
        Авторы: {currentAuthors}
      </p>
      <Link
        className={classes.card__link}
        to={`/${id}`}
      >
        Подробнее
      </Link>
    </article>
  );
};

export default Card;
