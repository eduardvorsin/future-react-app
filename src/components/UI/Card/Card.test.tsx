import React from 'react';
import { render, screen } from '@testing-library/react';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';
import Card from './Card';
import Image from '../../../assets/images/placeholder128x185-en.jpeg';
import i18n from '../../../localization/i18next';

describe('Card tests', () => {
  it('is rendered correctly', () => {
    const router = createMemoryRouter(
      [{
        path: '/',
        element: (
          <Card
            src={Image}
            id='test-card'
            title='test-card-title'
            alt='test-card-alt-text'
            maturityRating='NOT_MATURE'
          />
        ),
      }],
      { initialEntries: ['/'] },
    );
    render(<RouterProvider router={router} />);

    expect(screen.getByRole<HTMLDivElement>('article')).toBeInTheDocument();
    expect(screen.getByText<HTMLParagraphElement>(/authors/i)).toBeInTheDocument();
    expect(screen.getByText<HTMLParagraphElement>(/price/i)).toBeInTheDocument();
    expect(screen.getAllByRole<HTMLAnchorElement>('link')).toHaveLength(3);
  });

  it('if the authors prop is missing, it will output unknown in the corresponding paragraph', () => {
    const router = createMemoryRouter(
      [{
        path: '/',
        element: (
          <Card
            src={Image}
            id='test-card'
            title='test-card-title'
            alt='test-card-alt-text'
            maturityRating='NOT_MATURE'
          />
        ),
      }],
      { initialEntries: ['/'] },
    );
    render(<RouterProvider router={router} />);

    expect(screen.getByText<HTMLParagraphElement>(/unknown/i)).toBeInTheDocument();
  });

  it('if there is no price prop, it will output unknown in the corresponding paragraph', () => {
    const router = createMemoryRouter(
      [{
        path: '/',
        element: (
          <Card
            src={Image}
            id='test-card'
            title='test-card-title'
            alt='test-card-alt-text'
            maturityRating='NOT_MATURE'
          />
        ),
      }],
      { initialEntries: ['/'] },
    );
    render(<RouterProvider router={router} />);

    expect(screen.getByText<HTMLParagraphElement>(/not available/i)).toBeInTheDocument();
  });

  it('if maturityRating is equal to Mature, the card__img-wrapper--mature class is added to imageWrapper', () => {
    const router = createMemoryRouter(
      [{
        path: '/',
        element: (
          <Card
            src={Image}
            id='test-card'
            title='test-card-title'
            alt='test-card-alt-text'
            maturityRating='MATURE'
          />
        ),
      }],
      { initialEntries: ['/'] },
    );
    render(<RouterProvider router={router} />);

    expect(screen.getByTestId<HTMLDivElement>(/image-wrapper/i)).toHaveClass('card__img-wrapper--mature');
  });

  it('is a basic snapshot', () => {
    const router = createMemoryRouter(
      [{
        path: '/',
        element: (
          <Card
            src={Image}
            id='test-card'
            title='test-card-title'
            alt='test-card-alt-text'
            maturityRating='NOT_MATURE'
          />
        ),
      }],
      { initialEntries: ['/'] },
    );
    render(<RouterProvider router={router} />);

    expect(screen.getByRole<HTMLDivElement>('article')).toMatchSnapshot();
  });

  describe('Card integration tests', () => {
    it('the component with the text in Russian is displayed correctly', () => {
      i18n.changeLanguage('ru');

      const router = createMemoryRouter(
        [{
          path: '/',
          element: (
            <Card
              src={Image}
              id='test-card'
              title='test-card-title'
              alt='test-card-alt-text'
              maturityRating='NOT_MATURE'
            />
          ),
        }],
        { initialEntries: ['/'] },
      );
      render(<RouterProvider router={router} />);

      expect(screen.getByRole<HTMLDivElement>('article')).toBeInTheDocument();
      expect(screen.getByText<HTMLParagraphElement>(/авторы/i)).toBeInTheDocument();
      expect(screen.getByText<HTMLParagraphElement>(/цена/i)).toBeInTheDocument();
      expect(screen.getAllByRole<HTMLAnchorElement>('link')).toHaveLength(3);
    });

    it('if the current language is Russian and there is no prop authors, then it should output "неизвестно" in the corresponding paragraph', () => {
      i18n.changeLanguage('ru');

      const router = createMemoryRouter(
        [{
          path: '/',
          element: (
            <Card
              src={Image}
              id='test-card'
              title='test-card-title'
              alt='test-card-alt-text'
              maturityRating='NOT_MATURE'
            />
          ),
        }],
        { initialEntries: ['/'] },
      );
      render(<RouterProvider router={router} />);

      expect(screen.getByText<HTMLParagraphElement>(/неизвестно/i)).toBeInTheDocument();
    });

    it('If the current language is Russian and there is no prop price, then it should output "Нет в наличии" in the corresponding paragraph', () => {
      i18n.changeLanguage('ru');

      const router = createMemoryRouter(
        [{
          path: '/',
          element: (
            <Card
              src={Image}
              id='test-card'
              title='test-card-title'
              alt='test-card-alt-text'
              maturityRating='NOT_MATURE'
            />
          ),
        }],
        { initialEntries: ['/'] },
      );
      render(<RouterProvider router={router} />);

      expect(screen.getByText<HTMLParagraphElement>(/нет в наличии/i)).toBeInTheDocument();
    });
  });
});
