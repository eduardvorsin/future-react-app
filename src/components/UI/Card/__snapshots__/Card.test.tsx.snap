// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card tests is a basic snapshot 1`] = `
<article
  class=" card"
>
  <div
    class="card__img-wrapper "
    data-testid="image-wrapper"
  >
    <img
      alt="test-card-alt-text"
      class="card__img"
      height="185"
      src="test-file-stub"
      width="128"
    />
    <a
      href="/books/test-card"
    >
      Learn more
    </a>
  </div>
  <div
    class="card__content"
  >
    <p
      class="card__price"
    >
      <span
        class="sr-only"
      >
        Book price
        :
      </span>
      Not available
    </p>
    <h3
      class="title card__title title--dark title--level-6"
    >
      <a
        href="/books/test-card"
      >
        test-card-title
      </a>
    </h3>
    <p
      class="card__authors"
    >
      Authors
      : 
      Unknown
    </p>
  </div>
  <a
    aria-disabled="false"
    class="link card__more-link  link--secondary"
    href="/books/test-card"
  >
    Learn more
  </a>
</article>
`;
